Sources/GRPCHTTP2Core/Client/Connection/GRPCChannel.swift

```diff
@@ -0,0 +1,912 @@
+          self.past = [:]
+          self.queue = state.queue
+          self.nameResolverHandle = state.nameResolverHandle
+        }
+      }
+
+      struct Stopping {
+        /// Previously created load-balancers which are in the process of shutting down.
+        var past: [LoadBalancerID: LoadBalancer]
+
+        init(from state: Running) {
+          self.past = state.past
+        }
+
+        init(loadBalancers: [LoadBalancerID: LoadBalancer]) {
+          self.past = loadBalancers
+        }
+      }
+    }
+
+    /// The current state.
+    private var state: State
+    /// Whether the channel is running.
+    private var running: Bool
+
+    init() {
+      self.state = .notRunning(State.NotRunning())
+      self.running = false
+    }
+  }
+}
+
+@available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
+extension GRPCChannel.StateMachine {
+  mutating func start() {
+    precondition(!self.running, "channel must only be started once")
+    self.running = true
+  }
+
+  mutating func setNameResolverTaskHandle(_ handle: CancellableTaskHandle) {
+    switch self.state {
+    case .notRunning(var state):
+      state.nameResolverHandle = handle
+      self.state = .notRunning(state)
+    case .running, .stopping, .stopped, ._modifying:
+      fatalError("Invalid state")
+    }
+  }
+
+  enum LoadBalancerKind {
+    case roundRobin
+
+    func matches(loadBalancer: LoadBalancer) -> Bool {
+      switch (self, loadBalancer) {
+      case (.roundRobin, .roundRobin):
+        return true
+      }
+    }
+  }
+
+  enum OnChangeLoadBalancer {
+    case runLoadBalancer(LoadBalancer, stop: LoadBalancer?)
+    case updateLoadBalancer(LoadBalancer)
+    case none
+  }
+
+  mutating func changeLoadBalancerKind(
+    to newLoadBalancerKind: LoadBalancerKind,
+    _ makeLoadBalancer: () -> LoadBalancer
+  ) -> OnChangeLoadBalancer {
+    let onChangeLoadBalancer: OnChangeLoadBalancer
+
+    switch self.state {
+    case .notRunning(let state):
+      let loadBalancer = makeLoadBalancer()
+      let state = State.Running(from: state, loadBalancer: loadBalancer)
+      self.state = .running(state)
+      onChangeLoadBalancer = .runLoadBalancer(state.current, stop: nil)
+
+    case .running(var state):
+      self.state = ._modifying
+
+      if let next = state.next {
+        if newLoadBalancerKind.matches(loadBalancer: next) {
+          onChangeLoadBalancer = .updateLoadBalancer(next)
+        } else {
+          // The 'next' didn't become ready in time. Close it and replace it with a load-balancer
+          // of the next kind.
+          let nextNext = makeLoadBalancer()
+          let previous = state.next
+          state.next = nextNext
+          state.past[next.id] = next
+          onChangeLoadBalancer = .runLoadBalancer(nextNext, stop: previous)
+        }
+      } else {
+        if newLoadBalancerKind.matches(loadBalancer: state.current) {
+          onChangeLoadBalancer = .updateLoadBalancer(state.current)
```

Comment:
Isn't it possible that, even though the existing LB matches the new load balancer kind, the `makeLoadBalancer` closure builds an LB that's different in some other way (i.e. same kind but different configuration)? Wouldn't we want to update the LB here to the new one instead of using the current?

Reply:
The LB can only changed via the config from the resolver which will happen via `updateLoadBalancer`. The `makeLoadBalancer` closure mustn't change over time, all of the state it captures is immutable.